### DISCUSSION TITLE
Add site replication with https self signed certificate

### DISCUSSION THREAD

**Discussion by Ijoezl** [minio/minio#20551] [Q&A] [answered]
Hi,

Appreciate if you guys can help me with this. 

I want to configure mc admin replicate add to two of my minio node running https with self signed cert and face issue that the remote endpoint is offline. I am able to configure mc alias set to both minio nodes. I am running the mc command in minio168. The minio135 is always online and i never shutdown it.

I already add both public.crt in both VM .minio/certs/CAs and .mc/certs/CAs.

Kindly let me know if you guys need any additional info.

![mc admin replicate fail 2](https://github.com/user-attachments/assets/06e8f27c-502e-4679-b0a8-a01ccbf0ee77)

![mc admin info get through 1](https://github.com/user-attachments/assets/5f1079bf-e41e-4e52-b665-6e39236ec2cb)

![mc alias list 1](https://github.com/user-attachments/assets/a380d092-ec31-4bf1-b401-fd474f7746c9)

**Comment by marktheunissen** [comment]
If you login to a shell on 168, can you reach 135? Try something like a `curl -v https://10.10.6.135:9000` from the 168 node.

**Reply by Ijoezl** [reply]
Here is the result. Additional note, i generated my self signed cert using openssl.

![curl 135](https://github.com/user-attachments/assets/b80e572f-3bb7-44d9-b11c-f94e5a1a30ca)

If i run curl -kv i get this error instead. I guess it got through bypassing the cert verification but the API deny the request.

![curl kv 135](https://github.com/user-attachments/assets/3b36243c-586a-4ea5-b0fb-911aef4b520c)

**Reply by marktheunissen** [reply]
Add the self-signed certs to the trusted store, on Ubuntu something like:

`sudo cp selfsigned_cert.pem /usr/local/share/ca-certificates/self-signed-minio.crt`

`sudo update-ca-certificates`

Try the above, make sure curl has no errors with the certificate, restart minios and see if it works. 

Or use a cert signed by a public authority.

**Reply by Ijoezl** [reply]
But the Minio is being use as internal to test and it is Dev environment. I am not planning to tie it with a domain and use cert signed by public authority. Is there anyway to achieve this by not configuring domain to the both minio and test the site replication with https? 

From my understanding if using public authority cert its either letsencrypt which are not allowing me to configure cert with IP address and any paid SSL distributor and buy for a domain. Please advise if there is any other way to achieve this?

**Reply by marktheunissen** [reply]
Yes, please re-read my first comment, where I gave you the rough outline of how to recognize a self-signed cert:

```
Add the self-signed certs to the trusted store, on Ubuntu something like:

sudo cp selfsigned_cert.pem /usr/local/share/ca-certificates/self-signed-minio.crt

sudo update-ca-certificates

Try the above, make sure curl has no errors with the certificate, restart MinIO and see if it works.
```

**Reply by Ijoezl** [reply]
Since i am using RHEL for both VM. Here what i did to recognize the self-signed cert.

Copy both minio self-signed cert into /etc/pki/ca-trust/source/anchors/public.crt into both VM.

then run `update-ca-trust`

Then i verified the cert has been added by using `trust list` command

![trustedcert](https://github.com/user-attachments/assets/e8e5001c-3c87-42f8-a66f-e4005c0f59b7)

Im still unable to curl -v https://10.10.6.135:9000. Same error `SSL certificate problem: self-signed certificate`

The replication command error came out as:
Unable to add sites for replication. Error received when contacting a peer site (unable to fetch server info                                      for minio135: Get https://10.10.6.135:9000/minio/admin/v3/info?metrics=false: tls: failed to verify certificate: x509                                     : certificate signed by unknown authority)

**Reply by Ijoezl** [reply]
Is it due to that i did not include CA in my self-signed cert? 

i just created the private.key and the public.crt using openssl for the self-signed cert. 

Does CA needed for the trusted store to recognize the self-signed certificate?

**Reply by marktheunissen** [reply]
Yes you've obviously just missed something simple, recheck your steps and / or search for help on the internet with PKI infrastructure and follow the basic steps to get a CA and valid certificate recognized by Red Hat Linux. Make sure all servers and clients have the same basic trust store.

**Reply by Ijoezl** [reply]
After including the CA in my public.crt and recognize in trust store of the OS. I am able to curl -v https://10.10.6.135:9000 from 10.10.6.168 VM. 

![curl135ok](https://github.com/user-attachments/assets/a896c16b-db2a-4fc4-ae5d-cd8777dd625c)

But i am still unable to replicate with the error it cannot find the remote target endpoint https://10.10.6.135:9000.
While i am able to run mc ping to minio135 and mc admin info able to view the minio 135

![mc replicate issue latest](https://github.com/user-attachments/assets/a6145d89-4948-4810-b950-0428584aa6ae)

![mc ping and admin info ok](https://github.com/user-attachments/assets/4e799dd4-c5b1-4d55-83a1-0c6c2db76a00)

**Reply by Ijoezl** [reply, marked as answer]
I have fixed this by including the ca.crt inside the `--certs-dir` path of my minio. I have mislooked this and thought that the minio only needed public.crt to enable TLS.

Thanks for the help!
